[DumpRenderTree] Off-by-1 failure after running fonts on macOS

JonWBedard opened the pull request:
#### 1e8820b0c20a1b94809b4102358134d09e335508
<pre>
[DumpRenderTree] Off-by-1 failure after running fonts on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=273212">https://bugs.webkit.org/show_bug.cgi?id=273212</a>
<a href="https://rdar.apple.com/121476009">rdar://121476009</a>

Reviewed by Sam Sneddon.

If a ref test doesn&apos;t generate an image, the driver is in an undefined state. One possibility is that
we considered the driver &quot;finished&quot; before waiting for it&apos;s output. If that is true, we can possibly
cause every subsequent test run by this worker to fail.

* LayoutTests/platform/mac-wk1/TestExpectations: Skip only fonts/font-fallback-prefers-pictographs.html,
since that test occasionally crashes DumpRenderTree.
* LayoutTests/platform/mac-wk2/TestExpectations: Skip fast/forms/select/mac-wk2/inactive-appearance.html.
* Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
(SingleTestRunner._run_reftest): If a ref test doesn&apos;t generate any output, don&apos;t attempt to run the
baseline and kill the driver.
* Tools/Scripts/webkitpy/port/test.py:
(TestDriver.__init__): Add a &apos;is_valid_state&apos;, which is reset by &apos;stop&apos;.
(TestDriver.run_test): If a test is a ref test but doesn&apos;t produce an image, set &apos;is_valid_state&apos; to &apos;False&apos;.
This is an attempt to mock a bug in DumpRenderTree where a ref test returns an empty output initially and the
actual output on the subsequent test.
(TestDriver.stop): Reset &apos;is_valid_state&apos; to &apos;True&apos;.

Canonical link: <a href="https://commits.webkit.org/278180@main">https://commits.webkit.org/278180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f86fd5ffee6311683994bd59cd24d4d1c7ecd996

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49706 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28994 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52757 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52948 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/382 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52011 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35015 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26611 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40557 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51806 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26529 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21678 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/49562 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23972 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/44002 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8077 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45884 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44509 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54529 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24794 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20959 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47937 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26060 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46961 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10916 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26908 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25785 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
```diff
@@ -395,6 +395,17 @@ def _run_reftest(self):
         test_output.strip_stderror_patterns(self._port.stderr_patterns_to_strip())
 
+        if not test_output.image:
+            # The driver is misbehaiving, kill it so the error doesn't propigate to subsequent tests
```
```suggestion
            # The driver is misbehaving, kill it so the error doesn't propigate to subsequent tests
```

gsnedders reviewed (commented):
You said on Slack,

>  Turns out, this is actually covered by existing unit tests, which I know because my original implementation broke those tests.

Thus it's surprising that this hasn't visibly progressed any unit tests?

Also do we not still have a similar bug with `reference_output` below?

I also wonder if the Driver should deal with stopping itself when the input has `should_run_pixel_test` and it gets no image output?

Also the WK2 failure is surprising, given we haven't seen flakiness there AFAIK.

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
```diff
@@ -415,6 +426,10 @@ def _run_reftest(self):
 
+            if reference_output.image is None:
+                # The driver is misbehaving, kill it so the error doesn't propigate to subsequent tests
+                self._driver.stop()
```
Do we not want to return and/or break here out of the loop here?

JonWBedard replied:
I don't believe we do.

It's important to early return when it was the actual test which had an empty image, but less so when it's the reference, since we already know the test didn't have an empty image here, so we'll fail the comparison (which seems like the correct behavior)

gsnedders replied:
Ah, right; I'm forgetting how all of this fits together. So yeah, we already return ImageOnlyFailure as the result with `test_failures.FailureReftestNoImagesGenerated`, so it's fine to just let this fail.

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
```diff
@@ -415,6 +426,10 @@ def _run_reftest(self):
             reference_output.strip_stderror_patterns(self._port.stderr_patterns_to_strip())
 
+            if reference_output.image is None:
+                # The driver is misbehaving, kill it so the error doesn't propigate to subsequent tests
```
```suggestion
                # The driver is misbehaving, kill it so the error doesn't propagate to subsequent tests
```

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
```diff
@@ -395,6 +395,17 @@ def _run_reftest(self):
         test_output.strip_stderror_patterns(self._port.stderr_patterns_to_strip())
 
+        if test_output.image is None:
+            # The driver is misbehaving, kill it so the error doesn't propigate to subsequent tests
```
```suggestion
            # The driver is misbehaving, kill it so the error doesn't propagate to subsequent tests
```

JonWBedard committed:
[DumpRenderTree] Off-by-1 failure after running fonts on macOS
https://bugs.webkit.org/show_bug.cgi?id=273212
rdar://121476009

Reviewed by Sam Sneddon.

If a ref test doesn't generate an image, the driver is in an undefined state. One possibility is that
we considered the driver "finished" before waiting for it's output. If that is true, we can possibly
cause every subsequent test run by this worker to fail.

* LayoutTests/platform/mac-wk1/TestExpectations: Skip only fonts/font-fallback-prefers-pictographs.html,
since that test occasionally crashes DumpRenderTree.
* LayoutTests/platform/mac-wk2/TestExpectations: Skip fast/forms/select/mac-wk2/inactive-appearance.html.
* Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
(SingleTestRunner._run_reftest): If a ref test doesn't generate any output, don't attempt to run the
baseline and kill the driver.
* Tools/Scripts/webkitpy/port/test.py:
(TestDriver.__init__): Add a 'is_valid_state', which is reset by 'stop'.
(TestDriver.run_test): If a test is a ref test but doesn't produce an image, set 'is_valid_state' to 'False'.
This is an attempt to mock a bug in DumpRenderTree where a ref test returns an empty output initially and the
actual output on the subsequent test.
(TestDriver.stop): Reset 'is_valid_state' to 'True'.

Canonical link: https://commits.webkit.org/278180@main
